primitives/runtime/src/lib.rs

```diff
@@ -95,7 +95,52 @@ pub use either::Either;
 /// the block itself would allow swapping justifications to change the block's hash
 /// (and thus fork the chain). Sending a `Justification` alongside a block instead
 /// bypasses this problem.
-pub type Justification = Vec<u8>;
+pub type EncodedJustification = Vec<u8>;
+
+/// Justification together with an ID to tag it's origin.
+pub type Justification = (ConsensusEngineId, EncodedJustification);
+
+/// Collection of Justifications, since we might have more than one stored per block.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub struct Justifications(Vec<Justification>);
+
+impl Justifications {
+	/// Return an iterator over the Justifications.
+	pub fn iter(&self) -> impl Iterator<Item = &Justification> {
+		self.0.iter()
+	}
+
+	/// Append a justification.
+	pub fn push(&mut self, justification: Justification) {
```

Comment:
Or instead of a fancy map, we just call this one `append` and check for duplicate `justification.0` entries.

Reply:
Yeah that makes sense